flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -391,6 +396,10 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
                         () -> throwException(metadata, exception, producer),
                         "Failed to send data to Kafka");
             }
+
+            if (userCallback != null) {
+                userCallback.onCompletion(metadata, exception);
```

Comment:
Is this supposed to work with transactions? Does it make sense to forward uncommitted data?

You probably also want to go over the mailbox.

Reply:
There must be a `flush` before getting offsets from callback.

Reply:
Where is this method actually being implemented. I haven't figured out why this is important to a LogSink. Do you need it for deduplication?
How does it tie into the checkpointing approach of the KafkaWriter.

Reply:
Nowadays it is mainly used for `HybridSource`, table stores the offsets of the log corresponding to each snapshot generated (that is, obtained through this callback).
When Hybrid reading, it first reads the snapshot of the file store, then gets the offsets it recorded, and starts consuming the log store from that offsets.

Reply:
If this information is only needed on savepoint, then we should not call it writer callback and not invoke it outside of snapshot related operations.

I'd rather propose that Kafka internally stores the latest offset and relays that information on snapshot somewhere. If this is only used for `HybridSource`, we could also add a method to `Source` (or a specialized interface) to get the latest offset.

Reply:
This is used on checkpoint

Reply:
Writing
- write records...
- snapshot: get offsets

Reading
- read file store with a snapshot
- Then start reading the log store from the offsets corresponding to this snapshot

Reply:
If you mean to get the latest offsets via the Source API during the writing - snapshot phase, here's the problem:
Kafka does not guarantee that the latest offsets are always the latest, there may be messages still in the memory of the Kafka cluster, and the offsets you get may not include the offsets of messages in the memory.

Reply:
I created https://issues.apache.org/jira/browse/FLINK-25696 for this